kale/pipeline/mida_trainer.py

```diff
@@ -0,0 +1,201 @@
+"""
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.model_selection import GridSearchCV, LeaveOneGroupOut, ParameterGrid
+from sklearn.preprocessing import FunctionTransformer
+
+from ..embed.factorization import MIDA
+
+
+class MIDATrainer(BaseEstimator, ClassifierMixin):
+    """Trainer of pipeline: Transformer (optional) -> Maximum independence domain adaptation -> Estimator
+
+    Args:
+        estimator (BaseEstimator): Estimator object implementing 'fit'
+        estimator_param_grid (dict, optional): Grids for searching the optimal hyperparameters of the estimator.
+            Defaults to None.
+        transformer (BaseEstimator, optional): Transformer object implementing 'fit_transform'. Defaults to None.
+        mida_param_grid (dict, optional): Grids for searching the optimal hyperparameters of MIDA. Defaults to None.
+        transformer_param_grid (dict, optional): Grids for searching the optimal hyperparameters of the transformer.
+            Defaults to None.
+    """
+
+    def __init__(
+        self,
+        estimator,
+        estimator_param_grid=None,
+        transformer=None,
+        mida_param_grid=None,
+        transformer_param_grid=None,
+    ):
+        self.estimator = estimator
+        self.transformer = transformer
+        self.mida_param_grid = mida_param_grid
+        self.transformer_param_grid = transformer_param_grid
+        self.estimator_param_grid = estimator_param_grid
+        self.best_transformer = None
+        self.best_transformer_params = None
+        self.best_mida = None
+        self.best_mida_params = None
+        self.best_estimator = None
+        self.best_estimator_params = None
+        self.best_score = None
+
+    def fit(self, x, y, groups=None):
+        """Fit a Transformer -> Maximum independence domain adaptation -> Estimator pipeline with the given data x ,
+            labels y, and group labels groups
+
+        Args:
+            x (array-like): Training input data matrix, shape (n_samples, n_features)
+            y (array-like): Labels, shape (n_labeled_samples,)
+            groups (array-like, optional): Domain/group labels of x, shape (n_samples,)
+
+        Returns:
+            self
+        """
+        if self.transformer is None:
+            self.transformer = FunctionTransformer()
+        if self.transformer_param_grid is None:
+            self.transformer_param_grid = {key: [val] for key, val in self.transformer.get_params().items()}
+        transformer_grid = ParameterGrid(self.transformer_param_grid)
+        for transformer_params in transformer_grid:
+            self.transformer.set_params(**transformer_params)
+            x_transformed = self.transformer.fit_transform(x)
+            self._fit_mida(x_transformed, y, groups, transformer_params)
+
+        self.best_transformer = self.transformer.set_params(**self.best_transformer_params)
```

Comment:
I think cloning the original estimator with `sklearn.base.clone` first and then do `set_params` could be a better option to avoid messing with the original estimator.